cpu/kinetis_common/pwm.c

```diff
@@ -79,86 +126,164 @@ int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int re
+            #endif
+            #if PWM_1_CHANNELS > 5
+                PWM_1_PORT_CH5->PCR[PWM_1_PIN_CH5] = PORT_PCR_MUX(PWM_1_PIN_AF_CH5);
+            #endif
+            #if PWM_1_CHANNELS > 6
+                PWM_1_PORT_CH6->PCR[PWM_1_PIN_CH6] = PORT_PCR_MUX(PWM_1_PIN_AF_CH6);
+            #endif
+            #if PWM_1_CHANNELS > 7
+                PWM_1_PORT_CH7->PCR[PWM_1_PIN_CH7] = PORT_PCR_MUX(PWM_1_PIN_AF_CH7);
+            #endif
+            break;
+#endif
+
+        default:
+            return -1;
     }
 
-    /* cppcheck-suppress nullPointer */
-    tim->MODE = (1 << FTM_MODE_WPDIS_SHIFT);
+    /* disable write protect for changing settings */
+    tim->MODE = FTM_MODE_WPDIS_MASK;
 
-    /* setup pins, reset timer match value */
+    /* reset timer match value */
     for (int i = 0; i < channels; i++) {
-        port[i]->PCR[pins[i]] = PORT_PCR_MUX(af[i]);
-        /* cppcheck-suppress nullPointer */
         tim->CONTROLS[i].CnV = 0;
     }
 
     /* reset timer configuration registers */
-    /* cppcheck-suppress nullPointer */
     tim->COMBINE = 0;
+    tim->CNTIN = 0;
+    tim->SWOCTRL = 0;
 
     /* set prescale and mod registers to matching values for resolution and frequency */
-    if (resolution > 0xffff || (resolution * frequency) > pwm_clk) {
-        return -2;
-    }
-
-    /* cppcheck-suppress nullPointer */
-    tim->SC = FTM_SC_PS((pwm_clk / (resolution * frequency)) - 1);
-    /* cppcheck-suppress nullPointer */
-    tim->MOD = resolution;
+    tim->SC = FTM_SC_PS(prescaler);
+    tim->MOD = resolution - 1;
 
     /* set PWM mode */
+    uint32_t mode_mask;
     switch (mode) {
         case PWM_LEFT:
-            for (int i = 0; i < channels; i++) {
-                /* cppcheck-suppress nullPointer */
-                tim->CONTROLS[ftmchan[i]].CnSC = (1 << FTM_CnSC_MSB_SHIFT |
-                                                  1 << FTM_CnSC_ELSB_SHIFT);
-            }
-
+            mode_mask = (1 << FTM_CnSC_MSB_SHIFT) | (1 << FTM_CnSC_ELSB_SHIFT);
             break;
 
         case PWM_RIGHT:
-            for (int i = 0; i < channels; i++) {
-                /* cppcheck-suppress nullPointer */
-                tim->CONTROLS[ftmchan[i]].CnSC = (1 << FTM_CnSC_MSB_SHIFT |
-                                                  1 << FTM_CnSC_ELSA_SHIFT);
-            }
-
+            mode_mask = (1 << FTM_CnSC_MSB_SHIFT) | (1 << FTM_CnSC_ELSA_SHIFT);
             break;
 
         case PWM_CENTER:
-            for (int i = 0; i < channels; i++) {
-                /* cppcheck-suppress nullPointer */
-                tim->CONTROLS[ftmchan[i]].CnSC = (1 << FTM_CnSC_MSB_SHIFT);
-            }
-
-            /* cppcheck-suppress nullPointer */
+            mode_mask = (1 << FTM_CnSC_MSB_SHIFT);
             tim->SC |= (1 << FTM_SC_CPWMS_SHIFT);
             break;
     }
+    for (int i = 0; i < channels; i++) {
+        tim->CONTROLS[ftmchan[i]].CnSC = mode_mask;
```

Comment:
When testing #2992, I got a compiler warning that ```mode_mask``` may be uninitialized. Could you initialize ```mode_mask``` or add a default to switch-case?